### ISSUE TITLE
Logger doesn't allow plotting metrics against a different x-axis

### ISSUE THREAD

**Issue by ananyahjha93** [mosaicml/composer#1503]
** Environment **

- OS: Ubuntu 20.04
- Hardware (GPU, or instance type): 8xA100
- cuda: 11.3
- cudnn: 8
- pytorch: 1.12.1
- composer: dev branch installed from source
- deepspeed: 0.7.2
- transformers: 4.21.2

** To reproduce

I am trying to plot metrics passed to HuggingFaceModel against a different x-axis apart from having the same metrics being plotted against batch count. Using the following callback, I intend to plot metrics against FLOPs.

```
"""Plotting metrics against flops."""

from __future__ import annotations

import logging
from copy import deepcopy
from typing import Any, Callable, Dict, Optional, Union

import torch

from composer.core import State, Time
from composer.core.callback import Callback
from composer.core.time import TimeUnit
from composer.loggers import Logger

log = logging.getLogger(__name__)

__all__ = ['MetricAgainstFlops']


class MetricAgainstFlops(Callback):
    def __init__(
        self,
    ):
        pass
        # self.plot = plot
        # self.dataloader_label = dataloader_label

    def init(self, state: State, logger: Logger) -> None:
        self.num_params = sum(param.numel() for param in state.model.parameters())

        print('___________________________________________________________________________')
        print(state.model)
        print(f'Param count: {self.num_params / 2**20:.2f}M')
        print('___________________________________________________________________________')

    def _flops(self, num_tokens_trained_on: int) -> None:
        # this is specific to GPT/BLOOM class of transformer models
        return 6 * self.num_params * num_tokens_trained_on

    def _log_metrics(self, computed_metrics: Dict, state: State, logger: Logger) -> None:
        # this gives total tokens across all ranks
        # in eval loop, this gives number of tokens trained on until this eval loop
        num_tokens_trained_on = state.timestamp.token.value
        logger.log_metrics(
            {f'metrics/{state.dataloader_label}/{name}_vs_FLOPs': val for (name, val) in computed_metrics.items()},
            step=self._flops(num_tokens_trained_on)
        )

    def batch_end(self, state: State, logger: Logger) -> None:
        # train metrics have to plotted at batch_end
        metrics = deepcopy(state.train_metrics)

        computed_metrics = {}
        for metric_name, metric in metrics.items():
            computed_metrics[metric_name] = metric.compute()

        self._log_metrics(computed_metrics, state, logger)

    def eval_end(self, state: State, logger: Logger) -> None:
        # eval metrics have to be plotted at eval end
        metrics = deepcopy(state.eval_metrics[state.dataloader_label])

        computed_metrics = {}
        for metric_name, metric in metrics.items():
            computed_metrics[metric_name] = metric.compute()

        self._log_metrics(computed_metrics, state, logger)
```

## Expected behavior

Now, this should plot 4 separate train and 4 separate eval metrics; 2 against batches on the x-axis and 2 against FLOPs.

But instead in wandb, I am getting 4 train metrics plots but the metrics stop being plotted against batches after the first batch. And, in eval, I am only getting 2 metric plots.

<img width="2493" alt="Screen Shot 2022-09-05 at 8 32 57 PM" src="https://user-images.githubusercontent.com/7
<img width="1669" alt="Screen Shot 2022-09-05 at 8 32 48 PM" src="https://user-images.githubusercontent.com/7491256/188541488-5d17091b-c32d-46dd-9d60-fbd7ec8dedac.png">
491256/188541476-16b60226-6c19-4aeb-b6b4-709b8b164ed2.png">

I am not sure how my calling 

```
        logger.log_metrics(
            {f'metrics/{state.dataloader_label}/{name}_vs_FLOPs': val for (name, val) in computed_metrics.items()},
            step=self._flops(num_tokens_trained_on)
        )

```

in a callback prevents the trainer from calling its ``_compute_and_log_metrics()`` function correctly. @eracah Is this the same issue linked to: https://github.com/mosaicml/composer/issues/1473

**Comment by ananyahjha93** [comment]
This works in tensorboard.

<img width="1191" alt="Screen Shot 2022-09-05 at 8 55 28 PM" src="https://user-images.githubusercontent.com/7491256/188543799-49563a17-5d17-4b2e-9ed8-1a36034fb688.png">

**Comment by eracah** [comment]
Hi @ananyahjha93, so as we talked about on the community slack you want:
1. to plot various metrics vs. a custom x-axis, like flops
2. you want to do this fully programmatically (you don't want to use wandb's custom x-axis gui feature)
3. you want to use Tensorboard or WandB. Whichever one lets you get this functionality
4. Passing in flops instead of step to `log_metrics` works for tensorboard, but not for wandb.
5. You either want to make it work for WandB or if not formalize making it work for Tensorboard by making a callback

Did I capture everything?

**Comment by eracah** [comment]
Hi @ananyahjha93, so as we talked about on the community slack you want:
1. to plot various metrics vs. a custom x-axis, like flops
2. you want to do this fully programmatically (you don't want to use wandb's custom x-axis gui feature)
3. you want to use Tensorboard or WandB. Whichever one lets you get this functionality
4. Passing in flops instead of step to `log_metrics` works for tensorboard, but not for wandb.
5. You either want to make it work for WandB or if not formalize making it work for Tensorboard by making a callback

Did I capture everything?

**Comment by ananyahjha93** [comment]
@eracah yes, that's about it! we can expand the initial code I posted for this callback to let it select a metric or all the metrics in the composer model, take in a formula (ex- 6 * num_params * tokens for FLOPs), which accepts the state of the trainer and plots the required metrics

**Comment by eracah** [comment]
Ok, so that code works for Tensorboard right?

**Comment by eracah** [comment]
If it doesn't work with WandB, there is always the recommended way to do it in WandB, which is to [plot a table](https://docs.wandb.ai/ref/app/features/custom-charts). While we don't expose the `wandb.table` function through our logging interface yet, I think you could make a Callback that deals with this.

**Comment by ananyahjha93** [comment]
@eracah like we discussed, I'll work on the tensorboard callback for this!

**Comment by eracah** [comment]
Ok sounds good. Closing the issue. Native custom x-axis support will probably be a future feature, but for a now callback works. See this [link](https://docs.wandb.ai/guides/track/log#customize-axes-and-summaries-with-define_metric) for more info on custom wandb axes as well
